### PR TITLE
cdc: skip incremental scaned events after region fails

### LINKED ISSUE #17233: cdc: no more any events should be emit to a downstream after it retrieves an event error
### What version of TiKV are you using?
Master.

### What did happened?

For a given subscriped region, after emiting an `EventError` to one downstream, the downstream may still have an unfinished incremental scan, and the scan will still emit some events to the downstream.

This shouldn't be allowed. It

* can cause ticdc panic on "an committed entry with CommitTs less than ResolvedTs";
* may cause https://github.com/pingcap/tiflow/issues/11343.

### PR DESCRIPTION
**Pull request by hicqu**
### What is changed and how it works?

Issue Number: Close #17233 

What's Changed:

```commit-message
cdc: skip incremental scaned events after region fails
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/channel.rs [hunk 1]
```diff
@@ -244,22 +252,57 @@ impl From<MemoryQuotaExceeded> for SendError {
...
+    pub created: Instant,
+    pub event: CdcEvent,
+    pub size: usize,
+}
+
+pub struct ScanedEvent {
+    pub created: Instant,
+    pub event: CdcEvent,
+    pub size: usize,
+    pub truncated: Arc<AtomicBool>,
```
Comment (overvenus) [inline components/cdc/src/channel.rs:265]: Please add some comments.

File: components/cdc/src/delegate.rs [hunk 2]
```diff
@@ -136,6 +136,8 @@ pub struct Downstream {
     kv_api: ChangeDataRequestKvApi,
     filter_loop: bool,
     pub(crate) observed_range: ObservedRange,
+
+    pub(crate) scan_truncated: Arc<AtomicBool>,
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:140]: Please add some comments.

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: components/cdc/src/delegate.rs [hunk 3]
```diff
@@ -191,7 +197,14 @@ impl Downstream {
...
     }
 
+    /// EventErrors must be sent by this function. And we must ensure no more
+    /// events or ResolvedTs will be sent to the downstream after
+    /// `sink_error_event` is called.
     pub fn sink_error_event(&self, region_id: u64, err_event: EventError) -> Result<()> {
+        info!("cdc downstream meets region error";
+            "conn_id" => ?self.conn_id, "downstream_id" => ?self.id, "req_id" => self.req_id);
+
+        self.scan_truncated.store(true, Ordering::Release);
```
Comment (pingyu) [inline components/cdc/src/delegate.rs:207]: Will there be any race condition that some unexpected scaned entries have been send to TiCDC before the flag is set ?

File: components/cdc/src/channel.rs [hunk 4]
```diff
@@ -244,22 +252,57 @@ impl From<MemoryQuotaExceeded> for SendError {
     }
 }
 
+pub struct ObservedEvent {
+    pub created: Instant,
+    pub event: CdcEvent,
+    pub size: usize,
+}
+
+pub struct ScanedEvent {
```
Comment (pingyu) [inline components/cdc/src/channel.rs:261]: It looks to be not necessary to keep a `truncated` flag for every event. How about pass the flag to `Sink` or `Drain`.

File: components/cdc/src/channel.rs [hunk 5]
```diff
@@ -244,22 +252,57 @@ impl From<MemoryQuotaExceeded> for SendError {
...
-    unbounded_sender: UnboundedSender<(Instant, CdcEvent, usize)>,
-    bounded_sender: Sender<(Instant, CdcEvent, usize)>,
+    unbounded_sender: UnboundedSender<ObservedEvent>,
+    bounded_sender: Sender<ScanedEvent>,
     memory_quota: Arc<MemoryQuota>,
 }
 
 impl Sink {
+    /// Only observed events can be sent by `unbounded_send`.
     pub fn unbounded_send(&self, event: CdcEvent, force: bool) -> Result<(), SendError> {
```
Comment (pingyu) [inline components/cdc/src/channel.rs:298]: How about change `event` to `ObservedEvent` to make sure that we pass the correct events in. Ditto for `send_all`.

File: components/cdc/src/delegate.rs (hunk 3 above)
Comment (hicqu) [reply components/cdc/src/delegate.rs:207]: It's OK. We just need to ensure no more events will be sent to the downstream **after the `EventError` is drained**.

File: components/cdc/src/channel.rs (hunk 4 above)
Comment (hicqu) [reply components/cdc/src/channel.rs:261]: `Drain` is shared by many regions in one connection. If we keep it in `Drain`, a `HashMap` will be introduced.

**Review by pingyu — APPROVED** [review]
LGTM~

**Review by overvenus — APPROVED** [review]

**Pull request merged**